File: typescript/tracing/exa/_mock_server.ts
import { createServer, type IncomingMessage, type Server, type ServerResponse } from "node:http";

async function body(request: IncomingMessage): Promise<Record<string, unknown>> {
  const chunks: Buffer[] = [];
  for await (const chunk of request) chunks.push(Buffer.from(chunk));
  if (chunks.length === 0) return {};
  return JSON.parse(Buffer.concat(chunks).toString("utf8"));
}

function json(response: ServerResponse, status: number, payload: unknown): void {
  const encoded = JSON.stringify(payload);
  response.writeHead(status, {
    "content-type": "application/json",
    "content-length": Buffer.byteLength(encoded),
  });
  response.end(encoded);
}

function sse(response: ServerResponse, chunks: unknown[]): void {
  const encoded = `${chunks.map((chunk) => `data: ${JSON.stringify(chunk)}\n\n`).join("")}data: [DONE]\n\n`;
  response.writeHead(200, {
    "content-type": "text/event-stream",
    "content-length": Buffer.byteLength(encoded),
  });
  response.end(encoded);
}

function agentRun(status: "queued" | "completed", request?: unknown): Record<string, unknown> {
  return {
    id: "run-loopback",
    object: "agent.run",
    status,
    request,
    output: status === "completed" ? { text: "Deterministic agent result." } : undefined,
    usage: status === "completed" ? { agentComputeUnits: 1, searches: 1 } : undefined,
    costDollars:
      status === "completed"
        ? { total: 0.012, agentCompute: 0.005, search: 0.007 }
        : undefined,
  };
}

async function handle(request: IncomingMessage, response: ServerResponse): Promise<void> {
  const url = new URL(request.url ?? "/", "http://127.0.0.1");
  const payload = request.method === "POST" ? await body(request) : {};
  if (payload.query === "expected Exa provider error") {
    json(response, 429, { error: "deterministic Exa rate limit" });
    return;
  }

  if (request.method === "POST" && url.pathname === "/search") {
    if (payload.stream) {
      sse(response, [
        { choices: [{ delta: { content: "loopback " } }] },
        {
          choices: [{ delta: { content: "search" } }],
          citations: [
            {
              id: "source-1",
              url: "https://example.com/search",
              title: "Loopback source",
            },
          ],
        },
      ]);
      return;
    }
    json(response, 200, {
      results: [
        {
          id: "https://example.com/search",
          url: "https://example.com/search",
          title: "Loopback search result",
          text: "Deterministic Exa search content.",
          highlights: ["Deterministic Exa search content."],
        },
      ],
      requestId: "loopback-search-request",
      resolvedSearchType: payload.type ?? "auto",
      costDollars: { total: 0.007, search: { neural: 0.007 } },
    });
    return;
  }

  if (request.method === "POST" && url.pathname === "/contents") {
    const urls = Array.isArray(payload.urls) ? payload.urls : [];
    json(response, 200, {
      results: urls.map((item) => ({
        id: item,
        url: item,
        title: "Loopback page",
        text: "Deterministic page contents for instrumentation.",
        highlights: ["Deterministic page contents."],
      })),
      requestId: "loopback-contents-request",
      costDollars: { total: 0.001, contents: { text: 0.001 } },
    });
    return;
  }

  if (request.method === "POST" && url.pathname === "/answer") {
    if (payload.stream) {
      sse(response, [
        { choices: [{ delta: { content: "loopback " } }] },
        {
          choices: [{ delta: { content: "answer" } }],
          citations: [
            {
              id: "answer-source",
              url: "https://example.com/answer",
              title: "Answer source",
            },
          ],
        },
      ]);
      return;
    }
    json(response, 200, {
      answer: "A deterministic grounded answer.",
      citations: [
        {
          id: "answer-source",
          url: "https://example.com/answer",
          title: "Answer source",
        },
      ],
      costDollars: { total: 0.005 },
    });
    return;
  }

  if (request.method === "POST" && url.pathname === "/agent/runs") {
    if (request.headers.accept === "text/event-stream") {
      sse(response, [
        { id: "event-1", event: "run.started", data: { id: "run-loopback" } },
        {
          id: "event-2",
          event: "run.completed",
          data: { id: "run-loopback", output: { text: "agent result" } },
        },
      ]);
      return;
    }
    json(response, 200, agentRun("queued", payload));
    return;
  }

  if (request.method === "GET" && url.pathname === "/agent/runs/run-loopback") {
    json(response, 200, agentRun("completed"));
    return;
  }

  if (request.method === "POST" && url.pathname === "/research/v1") {
    json(response, 200, {
      researchId: "research-loopback",
      createdAt: Date.now(),
      model: payload.model ?? "exa-research-fast",
      instructions: payload.instructions ?? "loopback research",
      status: "pending",
    });
    return;
  }

  if (request.method === "GET" && url.pathname === "/research/v1/research-loopback") {
    json(response, 200, {
      researchId: "research-loopback",
      createdAt: Date.now(),
      model: "exa-research-fast",
      instructions: "Create a deterministic research brief.",
      status: "completed",
      output: { content: "Deterministic legacy research output." },
      costDollars: {
        total: 0.02,
        numPages: 1,
        numSearches: 1,
        reasoningTokens: 32,
      },
    });
    return;
  }

  json(response, 404, { error: `Unhandled ${request.method} ${url.pathname}` });
}

export interface MockExaServer {
  baseURL: string;
  close: () => Promise<void>;
}

export async function startMockExaServer(): Promise<MockExaServer> {
  const server: Server = createServer((request, response) => {
    void handle(request, response).catch((error) => {
      json(response, 500, { error: error instanceof Error ? error.message : String(error) });
    });
  });
  await new Promise<void>((resolve, reject) => {
    server.once("error", reject);
    server.listen(0, "127.0.0.1", () => resolve());
  });
  const address = server.address();
  if (!address || typeof address === "string") throw new Error("No mock Exa address");
  return {
    baseURL: `http://127.0.0.1:${address.port}`,
    close: () => new Promise<void>((resolve, reject) => server.close((error) => (error ? reject(error) : resolve()))),
  };
}

File: python/tracing/exa/_mock_server.py
from __future__ import annotations

import json
import threading
import time
from collections.abc import Iterator
from contextlib import contextmanager
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from typing import Any
from urllib.parse import urlparse


class _Handler(BaseHTTPRequestHandler):
    server_version = "RespanExaMock/1.0"

    def log_message(self, format: str, *args: Any) -> None:
        return

    def _body(self) -> dict[str, Any]:
        length = int(self.headers.get("content-length", "0"))
        if not length:
            return {}
        return json.loads(self.rfile.read(length).decode("utf-8"))

    def _json(self, status: int, payload: dict[str, Any]) -> None:
        body = json.dumps(payload).encode("utf-8")
        self.send_response(status)
        self.send_header("content-type", "application/json")
        self.send_header("content-length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def _sse(self, chunks: list[dict[str, Any]]) -> None:
        body = "".join(f"data: {json.dumps(chunk)}\n\n" for chunk in chunks)
        body += "data: [DONE]\n\n"
        encoded = body.encode("utf-8")
        self.send_response(200)
        self.send_header("content-type", "text/event-stream")
        self.send_header("content-length", str(len(encoded)))
        self.end_headers()
        self.wfile.write(encoded)

    def do_POST(self) -> None:
        path = urlparse(self.path).path
        body = self._body()
        if body.get("query") == "expected Exa provider error":
            self._json(429, {"error": "deterministic Exa rate limit"})
            return
        if path == "/search":
            if body.get("stream"):
                self._sse(
                    [
                        {"choices": [{"delta": {"content": "loopback "}}]},
                        {
                            "choices": [{"delta": {"content": "search"}}],
                            "citations": [
                                {
                                    "id": "source-1",
                                    "url": "https://example.com/search",
                                    "title": "Loopback source",
                                }
                            ],
                        },
                    ]
                )
                return
            self._json(
                200,
                {
                    "results": [
                        {
                            "id": "https://example.com/search",
                            "url": "https://example.com/search",
                            "title": "Loopback search result",
                            "text": "Deterministic Exa search content.",
                            "highlights": ["Deterministic Exa search content."],
                        }
                    ],
                    "requestId": "loopback-search-request",
                    "resolvedSearchType": body.get("type", "auto"),
                    "costDollars": {"total": 0.007, "search": {"neural": 0.007}},
                },
            )
            return
        if path == "/contents":
            urls = body.get("urls") or []
            self._json(
                200,
                {
                    "results": [
                        {
                            "id": url,
                            "url": url,
                            "title": "Loopback page",
                            "text": "Deterministic page contents for instrumentation.",
                            "highlights": ["Deterministic page contents."],
                        }
                        for url in urls
                    ],
                    "requestId": "loopback-contents-request",
                    "costDollars": {"total": 0.001, "contents": {"text": 0.001}},
                },
            )
            return
        if path == "/answer":
            if body.get("stream"):
                self._sse(
                    [
                        {"choices": [{"delta": {"content": "loopback "}}]},
                        {
                            "choices": [{"delta": {"content": "answer"}}],
                            "citations": [
                                {
                                    "id": "answer-source",
                                    "url": "https://example.com/answer",
                                    "title": "Answer source",
                                }
                            ],
                        },
                    ]
                )
                return
            self._json(
                200,
                {
                    "answer": "A deterministic grounded answer.",
                    "citations": [
                        {
                            "id": "answer-source",
                            "url": "https://example.com/answer",
                            "title": "Answer source",
                        }
                    ],
                    "costDollars": {"total": 0.005},
                },
            )
            return
        if path == "/agent/runs":
            if self.headers.get("accept") == "text/event-stream":
                events = [
                    {
                        "id": "event-1",
                        "event": "run.started",
                        "data": {"id": "run-loopback"},
                    },
                    {
                        "id": "event-2",
                        "event": "run.completed",
                        "data": {
                            "id": "run-loopback",
                            "output": {"text": "agent result"},
                        },
                    },
                ]
                encoded = "".join(
                    f"event: {event['event']}\ndata: {json.dumps(event)}\n\n"
                    for event in events
                ).encode("utf-8")
                self.send_response(200)
                self.send_header("content-type", "text/event-stream")
                self.send_header("content-length", str(len(encoded)))
                self.end_headers()
                self.wfile.write(encoded)
                return
            self._json(
                200,
                {
                    "id": "run-loopback",
                    "object": "agent.run",
                    "status": "queued",
                    "request": body,
                },
            )
            return
        if path == "/research/v1":
            self._json(
                200,
                {
                    "researchId": "research-loopback",
                    "createdAt": time.time() * 1000,
                    "model": body.get("model", "exa-research-fast"),
                    "instructions": body.get("instructions", "loopback research"),
                    "status": "pending",
                },
            )
            return
        self._json(404, {"error": f"Unhandled POST {path}"})

    def do_GET(self) -> None:
        path = urlparse(self.path).path
        if path == "/agent/runs/run-loopback":
            self._json(
                200,
                {
                    "id": "run-loopback",
                    "object": "agent.run",
                    "status": "completed",
                    "output": {"text": "Deterministic agent result."},
                    "usage": {"agentComputeUnits": 1.0, "searches": 1},
                    "costDollars": {
                        "total": 0.012,
                        "agentCompute": 0.005,
                        "search": 0.007,
                    },
                },
            )
            return
        if path == "/research/v1/research-loopback":
            now = time.time() * 1000
            self._json(
                200,
                {
                    "researchId": "research-loopback",
                    "createdAt": now,
                    "model": "exa-research-fast",
                    "instructions": "Create a deterministic research brief.",
                    "status": "completed",
                    "output": {"content": "Deterministic legacy research output."},
                    "costDollars": {
                        "total": 0.02,
                        "numPages": 1,
                        "numSearches": 1,
                        "reasoningTokens": 32,
                    },
                },
            )
            return
        self._json(404, {"error": f"Unhandled GET {path}"})


@contextmanager
def run_mock_exa_server() -> Iterator[str]:
    server = ThreadingHTTPServer(("127.0.0.1", 0), _Handler)
    thread = threading.Thread(target=server.serve_forever, daemon=True)
    thread.start()
    try:
        host, port = server.server_address
        yield f"http://{host}:{port}"
    finally:
        server.shutdown()
        thread.join(timeout=5)
        server.server_close()
